### PR TITLE
Add runner cli

### PR DESCRIPTION
**Pull request by cmarqu**
A "TODO" block is near the top, with a few other TODOs further below.

### CODE REVIEW

File: cocotb/runner.py [hunk 1]
```diff
@@ -934,13 +959,201 @@ def clean(recursive: bool = False) -> None:
...
+        help="Verilog include directories",
+    )
+    parser.add_argument(
+        "--library-name",
+        default="work",
+        help="Library name to compile into (default: %(default)s)",
+    )
+    parser.add_argument(
+        "--parameters",
+        "--generics",  # TODO: discuss this alt name
```
Comment (ktbarrett) [inline cocotb/runner.py:1032]: is there any way to support both?

File: cocotb/runner.py [hunk 2]
```diff
@@ -934,13 +959,201 @@ def clean(recursive: bool = False) -> None:
...
+        dest="parameters",
+        nargs="*",
+        metavar=("NAME=VALUE"),
+        action=ParseKwargs,
+        help="Verilog parameters or VHDL generics",
+    )
+    parser.add_argument(
+        "--plus-args",
+        "--plusargs",  # TODO: discuss this alt name
```
Comment (ktbarrett) [inline cocotb/runner.py:1041]: Much prefer `--plusargs`.

File: cocotb/runner.py [hunk 3]
```diff
@@ -934,13 +959,201 @@ def clean(recursive: bool = False) -> None:
...
+        dest="plus_args",
+        nargs="*",
+        default=[],
+        help="'plusargs' to set for the simulator",
+    )
+    parser.add_argument(
+        "--py-module",
```
Comment (ktbarrett) [inline cocotb/runner.py:1048]: I'd much prefer `--test-module` or just `--module`.

File: cocotb/runner.py (hunk 1 above)
Comment (cmarqu) [reply cocotb/runner.py:1143]: Both options work, or do you mean we should differentiate paramters from generics down to where the simulator option names are involved?

File: cocotb/runner.py (hunk 2 above)
Comment (cmarqu) [reply cocotb/runner.py:1152]: I agree, but lets move renaming to a different PR after this one is through.

File: cocotb/runner.py (hunk 1 above)
Comment (ktbarrett) [reply cocotb/runner.py:1143]: No, just leave them the alias rather than selecting one, whereas with the other, I would remove the original name.

File: cocotb/runner.py (hunk 2 above)
Comment (cmarqu) [reply cocotb/runner.py]: Done this anyway now.

**Comment by cmarqu** [comment]
I pushed again with name changes (most of which we discussed in the call). @themperek wanted to have these early to be able to convert more examples and test to use the runner and to write tests for the runner itself.

It would be good to get some more eyes on these names soon.

More work is to be done on the CLI part in the dict/list "grey area" of some arguments.

File: documentation/source/runner.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+See the full list of supported runner arguments by running ``cocotb-runner -h``.
+
+``cocotb-runner`` will set the return code ``1``
+if no tests have been run at all
+or if any failures have been found in the result files.
+It will return ``0`` otherwise.
+
+
+Usage with pytest
```
Comment (themperek) [inline documentation/source/runner.rst:44]: Maybe add that one could also run just do `python examples/simple_dff/test_dff.py` ?

File: cocotb/runner.py [hunk 5]
```diff
@@ -97,148 +163,196 @@ def test_command(self) -> Sequence[Command]:
 
     def build(
         self,
-        library_name: str = "work",
+        hdl_library: str = "work",
         verilog_sources: Sequence[PathLike] = [],
         vhdl_sources: Sequence[PathLike] = [],
         includes: Sequence[PathLike] = [],
-        defines: Sequence[str] = [],
+        defines: Mapping[str, object] = {},
```
Comment (themperek) [inline cocotb/runner.py:170]: Did we decide to use the list for defines (some bad experience from edalize)?

File: cocotb/runner.py [hunk 6]
```diff
@@ -2,16 +2,26 @@
 # Licensed under the Revised BSD License, see LICENSE for details.
 # SPDX-License-Identifier: BSD-3-Clause
 
+"""Build HDL and run cocotb tests."""
+
+# TODO: --clean cleans both build and test dir (if different). Do we also need --clean-build and --clean-test?
+# TODO: maybe do globbing and expanduser/expandvars in --include, --vhdl-sources, --verilog-sources
+# TODO: create a short README and a .gitignore (content: "*") in both build_dir and test_dir? (Some other tools do this.)
+# TODO: support timescale
+
```
Comment (themperek) [inline cocotb/runner.py:11]:
```suggestion
# TODO: support  custom dependencies

```

File: cocotb/runner.py [hunk 7]
```diff
@@ -97,148 +163,196 @@ def test_command(self) -> Sequence[Command]:
...
             os.remove(results_xml_file)
 
-        cmds = self.test_command()
+        cmds: Sequence[Command] = self.test_command()
         self.set_env()
-        self.execute(cmds, cwd=self.sim_dir)
+        self.execute(cmds, cwd=self.test_dir)
 
-        check_results_file(results_xml_file)
+        # only when running under pytest, raise exception with failing testcases
```
Comment (themperek) [inline cocotb/runner.py:334]: Why only for pytest?

File: cocotb/runner.py [hunk 8]
```diff
@@ -921,26 +1035,278 @@ def get_runner(simulator_name: str) -> Type[Simulator]:
...
+                key, value = value.split("=")
+            except ValueError:
+                raise SystemExit(
+                    f"ERROR: Expecting option format {option_string} [NAME=VALUE [NAME=VALUE ...]]"
+                )
+
+            getattr(namespace, self.dest)[key] = value
+
+
+def get_parser() -> argparse.ArgumentParser:
```
Comment (themperek) [inline cocotb/runner.py:1072]: Would need (at some point) to tests for this.

**Review by ktbarrett — COMMENTED** [review]
The gpi_interface should be a list of interfaces to load. The first one will be the entry (until I refactor all of that to enter into the GPI from a single point). No matter what the toplevel lang is, that order should be preserved, because it is possibly valid no matter the toplevel_lang. gpi_interface can be defaulted based on toplevel_lang if people find that easy. And a toplevel_lang of `None` should be valid which does not export the variable. cocotb.LANGUAGE is nonsense and really only useful for our regressions because it ultimately shouldn't matter to the user and with simulators that support multiple tops, it is possible to have multiple languages at the top.

File: cocotb/runner.py (hunk 5 above)
Comment (cmarqu) [reply cocotb/runner.py:170]: Yes, this is not reworked yet.

File: cocotb/runner.py (hunk 7 above)
Comment (cmarqu) [reply cocotb/runner.py]: Added some explanatory comment.

File: cocotb/runner.py (hunk 5 above)
Comment (cmarqu) [reply cocotb/runner.py:139]: Thought about this some more.

If we have this as a list, we would have to handle the format per simulator on our own in each of the tests and examples where we are using defines.
No such problem with a dict.

In case we *do* get the argument format wrong for edge cases, the user can just not use this option  and provide the settings in the correct simulator-specific format via `build_args`.

File: cocotb/runner.py [hunk 9]
```diff
@@ -97,148 +163,196 @@ def test_command(self) -> Sequence[Command]:
 
     def build(
         self,
-        library_name: str = "work",
+        hdl_library: str = "work",
```
[the remark this answers is not in the document]
Comment (cmarqu) [reply cocotb/runner.py]: Used `cocotb` now, but it's also confusing... is `cocotblib` any better? Other suggestions?

**Comment by cmarqu** [comment]
I'm not finding much time for this these days, sorry.

File: cocotb/runner.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ktbarrett) [reply cocotb/runner.py]: Perhaps `top`?

**Comment by themperek** [comment]
FYI: `HDL_TOPLEVEL_LANG` is not compatible with our ci and would require duplication in many places or some translation in the nox.

**Comment by cmarqu** [comment]
> Should we make most of those private https://cocotb--2897.org.readthedocs.build/en/2897/library_reference.html#api-runner ?

We could also make the whole runner private (`import _runner`) - the `cocotb-runner` command could stay the same and print a warning with each run.

File: cocotb/runner.py [hunk 10]
```diff
@@ -366,99 +475,91 @@ def test_command(self) -> List[Command]:
...
 
-        cmd = []
+        cmds = []
 
         if self.vhdl_sources:
-            cmd.append(["vlib", as_tcl_value(self.library_name)])
-            cmd.append(
+            cmds.append(["vlib", as_tcl_value(self.hdl_library)])
+            cmds.append(
                 ["vcom", "-mixedsvvh"]
```
Comment (ktbarrett) [inline cocotb/runner.py:538]: Not everyone has licenses for mixedsvvh mode. Also there are VHDL 2008 features that work in normal mode that don't work in mixedsvvh mode. This flag should only be added if a library has mixed sources.

Also this shows the issues with separate VHDL_SOURCES and VERILOG_SOURES. In mixed language designs the order of VHDL vs VERILOG sources matters, so they should not be considered separately.

**Review by ktbarrett — COMMENTED** [review]
I'm a little surprised that there aren't different subcommands for the cocotb-runner for building and running. I think this thing needs more serious work before we start recommending it.

**Comment by imphil** [comment]
Removing this from the 1.7 milestone to give us more time to address remaining issues.

File: cocotb/runner.py (hunk 10 above)
Comment (themperek) [reply cocotb/runner.py:538]: >Not everyone has licenses for mixedsvvh mode.

Addressed in https://github.com/cocotb/cocotb/pull/3103

>Also this shows the issues with separate VHDL_SOURCES and VERILOG_SOURES. In mixed language designs the order of VHDL vs VERILOG sources matters, so they should not be considered separately.

This should be possible. Multiple `build()` calls in given order. Do you have something else in mind.

File: cocotb/runner.py [hunk 11]
```diff
@@ -97,148 +132,208 @@ def test_command(self) -> Sequence[Command]:
 
     def build(
         self,
-        library_name: str = "work",
+        hdl_library: str = "cocotb",
```
[the remark this answers is not in the document]
Comment (themperek) [reply cocotb/runner.py:135]: Addressed in https://github.com/cocotb/cocotb/pull/3103

**Comment by imphil** [comment]
https://github.com/cocotb/cocotb/pull/3103 has been merged with some of the changes in here. 

I'm closing this PR for now, but it will stay around for when we want to get back to the CLI work.
